test/unit/assorted/server_discovery_and_monitoring.spec.test.ts

```diff
@@ -157,129 +167,109 @@ function cloneMap(map) {
-      testData.phases,
-      (phase, cb) => {
-        function phaseDone() {
-          if (phase.outcome) {
-            assertOutcomeExpectations(topology, events, phase.outcome);
-          }
+  await promisify(topology.connect.bind(topology))(options);
 
-          // remove error handler
-          topology.removeListener('error', incompatibilityHandler);
-          // reset the captured events for each phase
-          events = [];
-          cb();
-        }
+  for (const phase of testData.phases) {
+    const errorEvents = [];
+    topology.on('error', error => errorEvents.push(error));
 
-        const incompatibilityExpected = phase.outcome ? !phase.outcome.compatible : false;
-        if (incompatibilityExpected) {
-          topology.on('error', incompatibilityHandler);
+    if (phase.responses) {
+      for (const [address, hello] of phase.responses) {
+        topology.serverUpdateHandler(new ServerDescription(address, hello));
+      }
+      if (phase.outcome) {
+        assertOutcomeExpectations(topology, events, phase.outcome);
+        if (phase.outcome.compatible === false) {
+          for (const errorEvent of errorEvents) {
+            expect(errorEvent).to.be.instanceOf(MongoCompatibilityError);
+            expect(errorEvent.message).to.match(/but this version of the driver/);
+          }
         }
+      }
 
-        // if (phase.description) {
-        //   console.log(`[phase] ${phase.description}`);
-        // }
-
-        if (phase.responses) {
-          // simulate each hello response
-          phase.responses.forEach(response =>
-            topology.serverUpdateHandler(new ServerDescription(response[0], response[1]))
-          );
-          phaseDone();
-        } else if (phase.applicationErrors) {
-          eachAsyncSeries(
-            phase.applicationErrors,
-            (appError, phaseCb) => {
-              let withConnectionStub = sinon
-                .stub(ConnectionPool.prototype, 'withConnection')
-                .callsFake(withConnectionStubImpl(appError));
-
-              const server = topology.s.servers.get(appError.address);
-              server.command(ns('admin.$cmd'), { ping: 1 }, undefined, err => {
-                expect(err).to.exist;
-                withConnectionStub.restore();
-
-                phaseCb();
-              });
-            },
-            err => {
-              expect(err).to.not.exist;
-              phaseDone();
-            }
+      topology.removeAllListeners('error');
+      events = [];
+    } else if (phase.applicationErrors) {
+      for (const appError of phase.applicationErrors) {
+        const withConnectionStub = sinon
+          .stub(ConnectionPool.prototype, 'withConnection')
+          .callsFake(withConnectionStubImpl(appError));
+
+        const server = topology.s.servers.get(appError.address);
+        const res = promisify(server.command.bind(server))(ns('admin.$cmd'), { ping: 1 }, {});
+        withConnectionStub.restore();
```

Comment:
feels like the stub restore should happen later in the cleanup, or at least after we catch the thrownError in L249

Reply:
Moved it below awaiting the error, good call